services/visual-studio-marketplace/visual-studio-marketplace-downloads.service.js

```diff
@@ -46,23 +46,30 @@ module.exports = class VisualStudioMarketplaceDownloads extends VisualStudioMark
     ]
   }
 
-  async handle({ measure, extensionId }) {
-    const json = await this.fetch({ extensionId })
+  transform({ measure, json }) {
     const { statistics } = this.transformStatistics({ json })
     const { value: installs } = this.getStatistic({
       statistics,
       statisticName: 'install',
     })
 
+    // We already have the only data point for this badge type
+    // so no need to query for the value of the other data point.
     if (measure === 'i') {
-      return this.constructor.render({ measure, count: installs })
+      return { count: installs }
     }
 
-    const { value: updates } = this.getStatistic({
+    const { value: updateCount } = this.getStatistic({
       statistics,
       statisticName: 'updateCount',
     })
-    const downloads = +installs + +updates
-    return this.constructor.render({ measure, count: downloads })
+
+    return { count: updateCount + installs }
```

Comment:
I think the version of this I suggested is clearer because it makes the logic very clear. It's true this version saves a function call in a few cases, however that function is extraordinarily fast.

It could probably be even clearer if `transformStatistics` turned this into a normal object that could be destructured. Then the helper method wouldn't even be needed:

```js
const result = {}

extension.statistics.forEach(({ statisticName, value }) => {
  result[statisticName] = value
})

return { statistics: result }
```

Reply:
This is a pretty minor to change to what I had added in https://github.com/badges/shields/pull/2732 as part of the original refactor of this service, but certainly worth discussion/refactor (especially since some of the original complexity has been removed with Azure DevOps moved to a separate service)

The other downside with this current approach is that it involves repeated iterating through the statistic objects returned in that array 😉  That felt.. "dirty" to me so I started down a different path when I first started on the refactor, but ultimately I decided to stick with this implementation that was used in the LegacyService. I went that route since the Marketplace API in use is completely undocumented (as far as I can tell) and I actually felt that approach would be easier to grok for someone who wasn't intimately familiar with that API response and how Marketplace conditions change that response. 

The objects that exist in that statistic array can vary pretty wildly based on a variety of factors. From what I've seen, albeit anecdotally with a few extensions, there's not a ton of objects in that array (10-20 range max), but I have no idea what (if any) the upper limit is. However, when an extension has been installed at least once the `install` statistic has always been the first element in the array and the other two objects needed for the other badges are within the first ten elements. 

The other odd thing is that statistic objects aren't returned if a stat has a value of zero. To handle that scenario, all of the service implementations would need to have their own conditional assignment check:

`const installs = statistic.install ? statistic.install : 0;`

or in that base `transformStatistics` function we'd need to check all 3 of the potentially used values and ensure all three existed with a numeric value. 

What you're proposing (with the minor accompanying changes I described above) makes perfect sense to me, but I've done quite a bit of work with the API and have both published several extensions to the Marketplace. Do we think that approach (and _why_ behind some of those items) will be as clear to someone without that domain knowledge looking at this code a year+ from now? I could add some inline info although it would probably end up needing to be a short story (much like this comment 😆 )

Reply:
I really appreciate your thoroughness and thoughtfulness about this stuff! I did not realize there was special handling of 0; thanks for explaining that.

There are two ways I might suggest handling the issue you mention. One is destructuring with defaults:

```js
const { install = 0, update = 0 } = this.transform(...)
```

That would replace the missing values with zero.

Alternatively since the services are using a relatively short list of keys, the `transform` function could include a hard-coded safe list of keys, and return only those keys from `transform`, with zeros in place of missing values.

Part of what makes a good `transform` function is that it takes a sort of messy input and tidies it up into its useful form. Hopefully this is done with an implementation that is very safe and very obvious. That would steer me toward the har-coded safe list of keys, because it encapsulates the complexity. However if the list of keys is changing frequently, the first might be better because it just works, even if it breaks encapsulation slightly.

An advantage of the declarative Joi schema is that it documents quite a bit of the magic. Even if the schema changes later, at least there's one place you can look to see what kind of assumptions the code was making.

Another option is unit testing. You could make the transform method static, and add a couple `sazerac`-based tests. A couple sets of sample inputs and outputs – or honestly, even a single one – will do a nice job of documenting its function to someone else who might not know the API.

So paragraphs of documentation are fine, though a unit test beats a paragraph. 😄 

The transform code is fast enough that I wouldn't worry about the degrees of extra processing we're discussing. There probably is more overhead in the Joi validation, for example. I'd rather have easy to read and test badge code than opaque, prematurely optimized code.

None of this is blocking, really. Fine to consider this in a follow-on.

Reply:
> None of this is blocking, really. Fine to consider this in a follow-on.

I don't think this PR has anything urgent so I think I'd rather knock this out in this PR too (I'm ready to be done with the VS Marketplace badges 😄) 

>  Alternatively since the services are using a relatively short list of keys, the transform function could include a hard-coded safe list of keys, and return only those keys from transform, with zeros in place of missing values.

I'm thinking this will be the best approach too. It would only be three lines and a relatively short inline comment should be able to explain _why_ that's being done. I'll make this change tonight/tomorrow

> An advantage of the declarative Joi schema is that it documents quite a bit of the magic. Even if the schema changes later, at least there's one place you can look to see what kind of assumptions the code was making

I really wish the API response would always contain the core statistic key/value pair objects, with the value of zero when warranted. If that were the case this would've been a pretty basic schema. I'm not sure how feasible (or even possible) it would be to have Joi create an object in an array with a default value if that array doesn't contain an object with a certain field value... but I learn something new about Joi nearly every day ;)

Reply:
> I really wish the API response would always contain the core statistic key/value pair objects, with the value of zero when warranted. If that were the case this would've been a pretty basic schema. I'm not sure how feasible (or even possible) it would be to have Joi create an object in an array with a default value if that array doesn't contain an object with a certain field value... but I learn something new about Joi nearly every day 😉

You _could_ use Joi to filter the _post-transform_ object. i.e. turn the array into an object, then run it through a Joi schema that picks out the whitelisted values and sets the defaults. 😁

Reply:
Bingo! I added this to the `transformStatistics` function to ensure the defaults would be set.